### ISSUE TITLE
Odyssey Stats: accessibility audit

### ISSUE THREAD

**Issue by jeherve** [Automattic/wp-calypso#74321]
We recently received the following feedback about the updated Stats screen:

> The new statistics update has adversely impacted accessibility, as the dashboard lacks intuitiveness and contains multiple unlabeled buttons. As a screen reader user, I found it challenging to navigate.
-- [source](https://wptavern.com/jetpack-11-9-adds-sharing-button-for-mastodon-updates-stats-dashboard-design#comment-437420)

I'm no accessibility expert so this is hard for me to judge, but maybe we could review the buttons / links on the page, and ensure that they're all labeled.

Maybe we could also improve labeling on some of our more generic links (like "View all"), so they're more descriptive? This may help folks know what they're about to click on. Right now, I don't believe you can really know since the "View all" links are all the same without any clear indication of what "all" is referring to for each link.

Thanks!

**Comment by jeffgolenski** [comment]
This is great, thanks so much, Jer!

**Comment by monsieur-z** [comment]
I've taken a look at the _Traffic_ page in WPadmin. It's definitely not an audit nor an exhaustive list of what needs to be addressed, but here are some issues to fix to improve the page accessibility (screenshots from VoiceOver on maOS). I'm happy to provide more guidance about how to fix them if need be. I'm also not sure what criteria diagrams should respect, but I can dig deeper into that.

### Copy

- The page contains multiple links labeled `View details` and `Upgrade`. While that makes sense in context, screen reader users scanning all the links of the page may get confused by this. At least from a screen reader perspective, these links should have a unique label.
<img width="400" alt="Screenshot 2023-03-20 at 1 57 40 PM" src="https://user-images.githubusercontent.com/1620183/226426678-c55d058b-ee5b-4f79-993a-eaecc0238507.png">


### Markup
#### Duplicate `main` tag

- The page contains 2 `main` tags, which is confusing for screen reader users navigating by landmarks.
<img width="400" alt="Screenshot 2023-03-20 at 1 19 16 PM" src="https://user-images.githubusercontent.com/1620183/226418303-1f5e06fd-611e-42e5-9585-59c191d2125d.png">

#### Confusing outline

- The outline of the page (i.e. headings content and order) screen reader users see makes little sense when you compare it to the visual render of the page.
<img width="400" alt="Screenshot 2023-03-20 at 1 17 06 PM" src="https://user-images.githubusercontent.com/1620183/226418796-206be10e-f3fd-414e-b3d8-09a2b06f8968.png">

For the _Traffic_ page, the outline could be something like:
- `h1` Traffic (needs to be added)
   - `h2` 7-day highlights
   - `h2` Traffic data for [date interval]
     - `h3` Posts & pages
     - `h3` Referrers
     - ...
   - `h2` Enhance [site] with Jetpack Security, Performance, and Growth tools
     - `h3` Security
     - ...
   - `h2` Bring your stats with you using the Jetpack mobile app

#### Images alt text

- The product icon alternate text in the upgrade section (bottom of the page) is superfluous. Since the icon is followed by the name of the product, screen readers effectively announce the product name twice. Attribute should be empty: `alt=""`.
- `svg` tags in the mobile section are announced as images without any label. The first one could be ignored by screen readers (if we think it's purely decorative) while the second one should mention it's a QR code.


### Styling
#### Insufficient color contrast

- The contrast of the gray bits of text is too low. These bits should be a bit darker. Browser extensions like [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk?hl=en) or [axe DevTools](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd) will highlight which elements need to be addressed.

#### Focus state not visible

- Try to navigate through interactive elements by using the `Tab` key. You'll notice many elements on the page don't have a visible focus state. This prevents keyboard users from knowing the focus location on the page.


### Components

#### Tab nav

- [A common practice](https://inclusive-components.design/tabbed-interfaces/) is for the `Tab` key to move the focus from the selected tab to the first focusable element in the tab panel. Navigating between tab items should be possible using the arrow keys instead.

#### Highlights

- The use of a list tag around the cards could be beneficial here, so that screen reader users know beforehand how many items they're able to navigate.
- Given the current markup, screen readers announce a series of words with no clear relationship, e.g. `Visitors. 0. 0. 0%.`. I tested the page with empty stats, but I figure this should be something like `1000 visitors, compared to 500 visitors last week. A 100% increase.`.

#### Diagram

- Navigation arrow buttons have no labels. Screen reader users are not able to understand what they do.
- Stats cards below the diagram are interactive, but not focusable with a keyboard. They're also marked as links with no destination. They should be buttons.

**Comment by jeherve** [comment]
> I've taken a look at the Traffic page in WPadmin. 

@monsieur-z Thanks for taking a look! Was this for the Jetpack > settings > Traffic page? If so, it may make more sense to create a new issue in the Jetpack repo, where that codebase lives. 

We have some related issues we can x-link:

https://github.com/Automattic/jetpack/issues/6531

https://github.com/Automattic/jetpack/issues/6533

**Comment by monsieur-z** [comment]
Hey @jeherve! I checked the `Jetpack > Stats` page, specifically the _Traffic_ tab. Is it the one you wanted to test?

<img width="400" alt="Screenshot 2023-03-21 at 8 34 37 AM" src="https://user-images.githubusercontent.com/1620183/226607613-aca438eb-c291-4a7b-a3e7-3467db96ff23.png">

**Comment by jeherve** [comment]
> Is it the one you wanted to test?

Ah yes, perfect, thank you!

**Comment by jeffgolenski** [comment]
@monsieur-z Thanks so much for this. If you have time (only if), it would be great to get a similar audit of the Insights page as well. Then we can transfers all of these to issues and update stats in wp-admin and wordpress.com! cc @jsnmoon @keoshi @kangzj

**Comment by monsieur-z** [comment]
>  If you have time (only if), it would be great to get a similar audit of the Insights page as well. 
Sure thing! I'll take note to address that panel as well in the upcoming days.

**Comment by monsieur-z** [comment]
Here's a review of the _Insights_ tab on the _Jetpack > Stats_ page. Again, this is not exhaustive, but a good start. I'm available to help with fixes!

### Markup

#### Duplicate `main` tag
- This is the same issue as mentioned above.

#### Confusing outline
- This section only uses `h1` headings. There should be only one level 1 heading.

<img width="300" alt="Screenshot 2023-03-29 at 10 44 11 AM" src="https://user-images.githubusercontent.com/1620183/228584078-0efb863e-c479-4f0b-b10a-06db38be54a4.png">

The outline could be something like this:

- `h1` Insights (needs to be added)
  - `h2` 2023 in review
    - `h3` All-time stats
    - ...
  - `h2` All-time highlights
  - `h2` Posting activity
  - `h2` All-time insights
    - `h3` Total views
    - ...

### Styling

#### Insufficient color contrast
- The contrast between some text elements and their background is too low. Browser extensions like [Lighthouse](https://chrome.google.com/webstore/detail/lighthouse/blipmdconlkpinefehnmjammfjpmpbjk?hl=en) or [axe DevTools](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd) will highlight which elements need to be addressed.

#### Focus state not visible enough
- The focus outline around the toggle buttons could be slightly more obvious.

<img width="294" alt="Screenshot 2023-03-29 at 11 13 51 AM" src="https://user-images.githubusercontent.com/1620183/228585483-758a37da-6076-4536-be59-6052095cea5d.png">

#### Headings
- Headings of the same level should ideally look the same. `Posting Activity` is in a different typeface than the other level 2 headings.

<img width="260" alt="Screenshot 2023-03-29 at 11 17 01 AM" src="https://user-images.githubusercontent.com/1620183/228586420-6cd1886a-a8ff-4392-8496-d3d7947735ef.png">


### Components

#### Tab nav
- [A common practice](https://inclusive-components.design/tabbed-interfaces/) is for the Tab key to move the focus from the selected tab to the first focusable element in the tab panel. Navigating between tab items should be possible using the arrow keys instead. This is the same issue as the one described in the review of the _Traffic_ section.

#### 2023 in review
- Navigation arrow buttons have no labels. Screen reader users are not able to understand what they do.
- Cards should be part of a list, so that screen reader users know how many similar items to expect when navigating through the document.
- Screen readers announce stats this way: `Likes, 0, 0` (reading is also interrupted by groups opening and closing). It should read stats in a natural way, i.e., `0 likes`.

#### All-time highlights
- In the `All-time stats` card, stats should be marked up either as a list or as a table. This is to give screen reader users a heads-up of what's coming and an easier way to navigate.
- On small viewports, touch or mouse users can scroll the section horizontally to display more cards. However, keyboard users have no way to display the hidden content.
<img width="400" alt="Screenshot 2023-03-29 at 11 28 16 AM" src="https://user-images.githubusercontent.com/1620183/228589501-f7ff394f-1be9-4f67-9937-92c91687178c.png">

#### All-time insights
- Like the tab component, the arrow keys should allow navigating through the options of toggle buttons. This is currently done with the `Tab` key, which should move the focus to the next element instead.
- The stats in the `Comments` card look like a table (there's a header) but are not marked as such. It's confusing when reading the component with a screen reader. It should either include the stats in a `table` tag (and other cards do the same) or drop the header.
<img width="400" alt="Screenshot 2023-03-29 at 11 30 23 AM" src="https://user-images.githubusercontent.com/1620183/228590099-6f3babf1-463e-4daf-9048-19bc531a2f9e.png">

- When using the `Tab` key to navigate through stat lists, we can notice that 3 different elements under the same list item receive focus (the list item itself, which has the `role=button` attribute, and two separate links). However, these 3 elements have the exact same behavior. This makes navigating the list with a keyboard 3 times longer. The `li` shouldn't be interactive, and one of the two links should be removed.

**Comment by jeffgolenski** [comment]
@monsieur-z Thank you so much, Kevin!

**Comment by jeffgolenski** [comment]
@jsnmoon Design is currently working on a bunch of non-stats projects, but I'm wondering if you and your team can start queueing up some of these a11y issues into PRs? A lot of them don't require design.

**Comment by jsnmoon** [comment]
@jeffgolenski: Got it! It's on our maintenance board, so this is already prioritized 🙂

**Comment by jsnmoon** [comment]
I'm tracking these issues to fix for the traffic page:

### Copy

- [x] Add unique labels to links labeled as "View details" and "Upgrade" to avoid confusion for screen reader users. (#75732)

### Markup

- [ ] Remove duplicate `main` tags to avoid confusion for screen reader users navigating by landmarks.
- [x] Reorder headings to make the outline of the page more meaningful for screen reader users.
(https://github.com/Automattic/wp-calypso/pull/75919)
- [x] Add an `h2` tag with the text "Traffic" to provide a clear main heading for the page.
(https://github.com/Automattic/wp-calypso/pull/75919)
- [ ] Add proper alt text to the product icon in the upgrade section.
- [ ] Add proper labels to `svg` tags in the mobile section to ensure they are announced correctly.

### Styling

- [ ] Increase the contrast of the gray bits of text to meet accessibility standards.
- [ ] Ensure that interactive elements have a visible focus state for keyboard users.

### Components

- [ ] Make the tab navigation accessible using arrow keys instead of the `Tab` key.
- [ ] Use a list tag around the cards in the "Highlights" section to indicate the number of items and their relationship to each other.
- [ ] Update the text in the "Highlights" section to provide meaningful information for screen reader users.
- [ ] Add labels to navigation arrow buttons in the "Diagram" section to help screen reader users understand their function.
- [ ] Make the stats cards in the "Diagram" section focusable with a keyboard and mark them as buttons instead of links.

**Comment by annacmc** [comment]
I'm tracking the following issues for the insights page fix:

## Markup

- [ ] Remove duplicate main tags to avoid confusion for screen reader users navigating by landmarks. https://github.com/Automattic/wp-calypso/pull/75991
- [x] Restructure heading hierarchy with one h1 heading and appropriate subheadings. https://github.com/Automattic/wp-calypso/pull/75802

## Styling

- [ ] Improve color contrast between text elements and their backgrounds.
- [ ] Enhance focus state visibility for toggle buttons.
- [ ] Standardize styling for same-level headings.

## Components 

- [ ] Implement Tab and arrow key navigation for tab components.
- [ ] Add labels to navigation arrow buttons.
- [ ] Organize cards as lists or tables for better screen reader navigation.
- [ ] Improve screen reader announcement of stats.
- [ ] Enable horizontal scrolling for keyboard users on small viewports.
- [ ] Allow arrow key navigation for toggle buttons.
- [ ] Markup stats in the Comments card as a table or remove the header.
- [ ] Simplify keyboard navigation for stat lists by removing redundant interactive elements.
